### DISCUSSION TITLE
allow upscaling of retreived artwork

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#43] [Ideas]
as reference to my post on coreelec forum:
https://discourse.coreelec.org/t/graphical-front-panel-display/12932/210

Artwork do not get upscaled (only downscaled)  
This could be problematic on display with high pixel density as the displayed size is limited by the resilution of the thumb retreived from kodi.

Of course upscaling will decrease quality but it should stay acceptable.
I personally plan to use a 12/15" display with 1920x1080 resolution.

Is there anyway to upscale the retrieved picture?

**Comment by mattblovell** [comment]
Upsizing would need to use ``resize()`` rather than the current ``thumbnail()`` code:

[Pillow Image.resize](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.resize)

Back in the commit history, kodi_display_panel.py originally used ``resize()``, but care was needed to ensure that the existing aspect ratio was maintained. The ``thumbnail()`` function takes care of that automatically, but only wants to go one direction – smaller.

All of the resampling filters listed for both ``resize()`` and ``thumbnail()`` are concerned with the mapping of multiple input pixels to a single output pixel (i.e., downsizing):

[Pillow Filters ](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#concept-filters)

So, I don't know if upsizing is actually possible or not and, if possible, I don't know what the result would like look.  The line I've often heard is that resizing cannot "fill in" information that isn't present; the result may therefore just look blocky.

**Comment by mattblovell** [comment]
The "Resizing Images" section of this page

https://auth0.com/blog/image-processing-in-python-with-pillow/#Resizing-Images

states the following:

> Another significant difference between the ``resize()`` and ``thumbnail()`` functions is that the ``resize()`` function 'blows up' an image if given parameters that are larger than the original image, while the ``thumbnail()`` function doesn't. For example, given an image of size 400x200, a call to ``resize((1200, 600))`` will create a larger-sized image 1200x600; thus, the image will have lost some definition and is likely to be blurry compared to the original. On the other hand, a call to ``thumbnail((1200, 600))`` using the original image will result in an image that keeps its size 400x200 since both the width and height are less than the specified maximum width and height.

Note the "[larger image] is likely to be blurry compared to the original" description.

**Comment by mattblovell** [comment]
I'll look at digging up the older code that used ``resize()`` and adding it to the relatively-new ``paste_artwork()`` function.  

Enlarging will likely be qualified with an additional key (mutually exclusive with the existing ``center_sm`` functionality).

**Comment by mattblovell** [comment]
Should be possible in kodi_panel v1.41

**Comment by nico1080** [comment]
What is the correct syntax; in my TOML file I have the following

> [shared_element.thumb]   # artwork / movie poster
posx = 0
posy = 0
width  = 1000
height = 1000
#size = 900
center_sm = 1
display_ifnot = [ "VideoPlayer.Cover", "DefaultVideoCover.png"]
enlarge = 1

It does not enlarge the picture.
I also tried to specify `size` instead of `width` and `height` item same result, artwork get displayed to ist full size with out upscaling

**Comment by mattblovell** [comment]
Please try kodi_panel v1.48.  There was a bug that was introduced into ``get_artwork()`` when I was separating the handling of AirPlay cover art and introducing the new callback functions for art.

> What is the correct syntax;

For *audio* artwork going through a ``thumb`` element or making use of the ``element_audio_cover()`` callback, only ``size`` is expected (as audio covers are generally square).  I did just introduce code to check for ``width`` and ``height`` if ``size`` is not present.

All video covers and the ``element_generic_artwork()`` callback **only** check for ``width`` and ``height``.

Finally, only one of ``center_sm`` or ``enlarge`` needs to be specified.  They are mutually-exclusive options (since, if ``enlarge`` is specified, then ``center_sm`` never has a reason to get examined).

**Comment by nico1080** [comment]
I confirm it work with 1.48. (both audio with `size `and video with `width `and `height `)

Thanks
